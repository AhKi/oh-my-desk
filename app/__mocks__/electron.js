export default jest.fn();

export const webContents = {
  getAllWebContents: jest.fn(() => []),
};

export const ipcMain = {
  on: jest.fn(),
};

export const BrowserWindow = jest.fn(info => ({
  close: jest.fn(),
  loadURL: jest.fn(),
  once: jest.fn(),
  on: jest.fn(),
  setAlwaysOnTop: jest.fn(),
  show: jest.fn(),
  showInactive: jest.fn(),
  getContentBounds: jest.fn(() => ({
    x: info.x,
    y: info.y,
    height: info.height,
    width: info.width,
  })),
  webContents: {
    on: jest.fn(),
    send: jest.fn(),
    openDevTools: jest.fn(),
  },
}));

export const ipcRenderer = {
  on: jest.fn(),
  send: jest.fn(),
};

export const remote = {
  getGlobal: jest.fn(),
  getCurrentWindow: jest.fn(),
  Menu: {
    buildFromTemplate: jest.fn(() => ({
      popup: jest.fn(),
    })),
  },
  app: {
    getVersion: jest.fn(),
  },
};

export const shell = {
  openExternal: jest.fn(),
};

export const app = {
  getName: jest.fn(() => 'oh-my-desk'),
  getPath: jest.fn(arg => arg),
  getVersion: jest.fn(),
  getAppPath: jest.fn(() => 'mock-path'),
  isReady: jest.fn(),
  on: jest.fn(),
};

export const Menu = {
  buildFromTemplate: jest.fn(() => ({
    popup: jest.fn(),
  })),
  setApplicationMenu: jest.fn(),
};

export const globalShortcut = {
  register: jest.fn(),
  unregister: jest.fn(),
};
