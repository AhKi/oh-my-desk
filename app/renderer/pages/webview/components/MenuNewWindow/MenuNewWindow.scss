.MenuNewWindow {
  background-color: gainsboro;
  position: absolute;
}

.MenuNewWindow__btn {
  width: 100%;
  height: 100%;
}

.MenuNewWindow__list {
  width: 100px;
  height: 30px;
  background-color: antiquewhite;
  border: 1px solid black;
  margin-bottom: -1px;
}
