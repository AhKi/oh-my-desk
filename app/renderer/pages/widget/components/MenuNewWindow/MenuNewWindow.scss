.MenuNewWindow {
  background-color: #414141;
  color: $color-white-opacity50;
  border-radius: 5px;

  position: absolute;
}

.MenuNewWindow__btn {
  color: inherit;
  width: 100%;
  height: 100%;
}

.MenuNewWindow__list {
  padding: 10px;
  color: inherit;
  font-size: 13px;
  font-weight: 500;
  height: 35px;
  border-bottom: 1px solid #7D7D7D;
  margin-bottom: -1px;

  &:hover {
    background-color: #515151;
    border-radius: 5px;
    color: $color-white;
  }
}

.MenuNewWindow__list--bottom {
  border: 0;
}
