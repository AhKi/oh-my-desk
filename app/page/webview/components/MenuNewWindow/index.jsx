import React from 'react';
import PropTypes from 'prop-types';
import { shell } from 'electron';
import i18n from 'constants/i18n';
import './MenuNewWindow.scss';

const propTypes = {
  url: PropTypes.string,
  webview: PropTypes.object, // eslint-disable-line react/forbid-prop-types
  x: PropTypes.number,
  y: PropTypes.number,
  onClose: PropTypes.func,
};

const defaultProps = {
  url: '',
  webview: null,
  x: 0,
  y: 0,
  onClose() {},
};

class MenuNewWindow extends React.Component {
  constructor(props) {
    super(props);
    this.handleClose = this.handleClose.bind(this);
    this.handleOpenBrowser = this.handleOpenBrowser.bind(this);
    this.handleOpenWidget = this.handleOpenWidget.bind(this);
    this.handleMovementPage = this.handleMovementPage.bind(this);
  }

  componentDidMount() {
    window.addEventListener('mouseup', this.handleClose);
  }

  componentWillUnmount() {
    window.removeEventListener('mouseup', this.handleClose);
  }

  handleClose() {
    const { onClose } = this.props;
    onClose();
  }

  handleOpenBrowser() {
    const { url } = this.props;
    shell.openExternal(url);
  }

  handleMovementPage() {
    const { webview, url } = this.props;
    webview.loadURL(url);
  }

  handleOpenWidget() {
    const { url } = this.props;
    console.log('url: ', url);
  }

  render() {
    const {
      x,
      y,
    } = this.props;
    const text = i18n().contextMenu;

    return (
      <ul
        className="MenuNewWindow"
        style={{ left: x, top: y }}
      >
        <li className="MenuNewWindow__list">
          <button
            className="MenuNewWindow__btn"
            type="button"
            onMouseUp={this.handleOpenBrowser}
          >
            {text.openNewBrowser}
          </button>
        </li>
        <li className="MenuNewWindow__list">
          <button
            className="MenuNewWindow__btn"
            type="button"
            onMouseUp={this.handleMovementPage}
          >
            {text.moveBrowser}
          </button>
        </li>
        <li className="MenuNewWindow__list">
          <button
            className="MenuNewWindow__btn"
            type="button"
            onMouseUp={this.handleOpenWidget}
          >
            {text.openNewWidget}
          </button>
        </li>
      </ul>
    );
  }
}

MenuNewWindow.propTypes = propTypes;
MenuNewWindow.defaultProps = defaultProps;

export default MenuNewWindow;
