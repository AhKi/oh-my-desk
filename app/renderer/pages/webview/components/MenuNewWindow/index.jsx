import React from 'react';
import PropTypes from 'prop-types';
import { shell } from 'electron';
import i18n from 'constants/i18n';
import './MenuNewWindow.scss';

const propTypes = {
  url: PropTypes.string,
  webview: PropTypes.object, // eslint-disable-line react/forbid-prop-types
  x: PropTypes.number,
  y: PropTypes.number,
  onClose: PropTypes.func,
  onMakeWidget: PropTypes.func,
};

const defaultProps = {
  url: '',
  webview: null,
  x: 0,
  y: 0,
  onClose() {},
  onMakeWidget() {},
};

class MenuNewWindow extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      x: props.x,
      y: props.y,
    };
    this.menuRef = React.createRef();
    this.handleClose = this.handleClose.bind(this);
    this.handleOpenBrowser = this.handleOpenBrowser.bind(this);
    this.handleOpenWidget = this.handleOpenWidget.bind(this);
    this.handleMovementPage = this.handleMovementPage.bind(this);
  }

  componentDidMount() {
    const { x, y, webview } = this.props;
    const screenWidth = webview.clientWidth;
    const screenHeight = webview.clientHeight;
    const menuWidth = this.menuRef.current.clientWidth;
    const menuHeight = this.menuRef.current.clientHeight;

    this.setState({
      x: x + menuWidth > screenWidth ? x - menuWidth : x,
      y: y + menuHeight > screenHeight ? y - menuHeight : y,
    });

    window.addEventListener('mouseup', this.handleClose);
    window.addEventListener('resize', this.handleClose);
  }

  componentWillUnmount() {
    window.removeEventListener('mouseup', this.handleClose);
    window.removeEventListener('resize', this.handleClose);
  }

  handleClose() {
    const { onClose } = this.props;
    onClose();
  }

  handleOpenBrowser() {
    const { url } = this.props;
    shell.openExternal(url);
  }

  handleMovementPage() {
    const { webview, url } = this.props;
    webview.loadURL(url);
  }

  handleOpenWidget() {
    const { url, onMakeWidget } = this.props;

    onMakeWidget({ url });
  }

  render() {
    const {
      x,
      y,
    } = this.state;
    const text = i18n().contextMenu;

    return (
      <ul
        className="MenuNewWindow"
        ref={this.menuRef}
        style={{ left: x, top: y }}
      >
        <li className="MenuNewWindow__list">
          <button
            className="MenuNewWindow__btn"
            type="button"
            onMouseUp={this.handleOpenBrowser}
          >
            {text.openNewBrowser}
          </button>
        </li>
        <li className="MenuNewWindow__list">
          <button
            className="MenuNewWindow__btn"
            type="button"
            onMouseUp={this.handleMovementPage}
          >
            {text.moveBrowser}
          </button>
        </li>
        <li className="MenuNewWindow__list">
          <button
            className="MenuNewWindow__btn"
            type="button"
            onMouseUp={this.handleOpenWidget}
          >
            {text.openNewWidget}
          </button>
        </li>
      </ul>
    );
  }
}

MenuNewWindow.propTypes = propTypes;
MenuNewWindow.defaultProps = defaultProps;

export default MenuNewWindow;
