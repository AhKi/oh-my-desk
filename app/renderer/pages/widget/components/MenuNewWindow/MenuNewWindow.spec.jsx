import React from 'react';
import { mount } from 'enzyme';
import { shell } from 'electron';
import MenuNewWindow from '.';

describe('<MenuNewWindow />', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should match to snapshot when render default', () => {
    const wrapper = mount(<MenuNewWindow widget={{}} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should test componentDidMount', () => {
    const componentDidMount = jest.spyOn(MenuNewWindow.prototype, 'componentDidMount');
    const componentWillUnmount = jest.spyOn(MenuNewWindow.prototype, 'componentWillUnmount');
    window.addEventListener = jest.fn();
    window.removeEventListener = jest.fn();
    const wrapper = mount(<MenuNewWindow widget={{}} />);
    const { handleClose } = wrapper.instance();

    expect(componentDidMount).toHaveBeenCalledTimes(1);
    expect(window.addEventListener).toHaveBeenCalledWith(
      'mouseup',
      handleClose,
    );
    expect(window.addEventListener).toHaveBeenCalledWith(
      'resize',
      handleClose,
    );

    wrapper.unmount();
    expect(componentWillUnmount).toHaveBeenCalledTimes(1);
    expect(window.removeEventListener).toHaveBeenCalledWith(
      'mouseup',
      handleClose,
    );
    expect(window.removeEventListener).toHaveBeenCalledWith(
      'resize',
      handleClose,
    );
  });

  it('should call handleClose', () => {
    const onClose = jest.fn();
    const wrapper = mount(<MenuNewWindow widget={{}} onClose={onClose} />);

    wrapper.instance().handleClose();

    expect(onClose).toHaveBeenCalledTimes(1);
    expect(onClose).toHaveBeenCalledWith();
  });

  it('should call handleOpenBrowser', () => {
    const wrapper = mount(
      <MenuNewWindow
        url="mock-url"
        widget={{}}
      />,
    );

    wrapper.instance().handleOpenBrowser();

    expect(shell.openExternal).toHaveBeenCalledTimes(1);
    expect(shell.openExternal).toHaveBeenCalledWith('mock-url');
  });

  it('should call handleMovementPage', () => {
    const loadURL = jest.fn();
    const widget = {
      loadURL,
    };
    const wrapper = mount(
      <MenuNewWindow
        url="mock-url"
        widget={widget}
      />,
    );

    wrapper.instance().handleMovementPage();

    expect(loadURL).toHaveBeenCalledTimes(1);
    expect(loadURL).toHaveBeenCalledWith('mock-url');
  });
});
