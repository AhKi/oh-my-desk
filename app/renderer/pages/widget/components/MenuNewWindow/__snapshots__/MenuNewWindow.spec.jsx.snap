// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MenuNewWindow /> should match to snapshot when render default 1`] = `
<MenuNewWindow
  onClose={[Function]}
  onMakeWidget={[Function]}
  url=""
  widget={Object {}}
  x={0}
  y={0}
>
  <ul
    className="MenuNewWindow"
    style={
      Object {
        "left": 0,
        "top": 0,
      }
    }
  >
    <li
      className="MenuNewWindow__list"
    >
      <button
        className="MenuNewWindow__btn"
        onMouseUp={[Function]}
        type="button"
      >
        Open with a new browser
      </button>
    </li>
    <li
      className="MenuNewWindow__list"
    >
      <button
        className="MenuNewWindow__btn"
        onMouseUp={[Function]}
        type="button"
      >
        Move
      </button>
    </li>
    <li
      className="MenuNewWindow__list MenuNewWindow__list--bottom"
    >
      <button
        className="MenuNewWindow__btn"
        onMouseUp={[Function]}
        type="button"
      >
        Open with a new widget
      </button>
    </li>
  </ul>
</MenuNewWindow>
`;
